### PR TITLE
fix: python 2 Function interfaces recompatibility

### LINKED ISSUE #2060: Function interface crash when upgrading nipype
### Summary
The function interface crashes, in a code that was working in an older nipype installation (version details below). From the error, it seems I am doing something wrong when importing other functions inside of the function interface, but I don't know what it is exactly or how to fix it.

The function itself is intended to convert html into pdf and send it as an email attachment.

### Actual behavior
The function below does not run, and this is the crash information:
```
170601-20:23:45,389 workflow INFO:
	 Traceback (most recent call last):
  File "/apps/software/anaconda/4.3.1/lib/python2.7/site-packages/nipype/pipeline/plugins/multiproc.py", line 52, in run_node
    result['result'] = node.run(updatehash=updatehash)
  File "/apps/software/anaconda/4.3.1/lib/python2.7/site-packages/nipype/pipeline/engine/nodes.py", line 372, in run
    self._run_interface()
  File "/apps/software/anaconda/4.3.1/lib/python2.7/site-packages/nipype/pipeline/engine/nodes.py", line 482, in _run_interface
    self._result = self._run_command(execute)
  File "/apps/software/anaconda/4.3.1/lib/python2.7/site-packages/nipype/pipeline/engine/nodes.py", line 613, in _run_command
    result = self._interface.run()
  File "/apps/software/anaconda/4.3.1/lib/python2.7/site-packages/nipype/interfaces/base.py", line 1081, in run
    runtime = self._run_wrapper(runtime)
  File "/apps/software/anaconda/4.3.1/lib/python2.7/site-packages/nipype/interfaces/base.py", line 1029, in _run_wrapper
    runtime = self._run_interface(runtime)
  File "/apps/software/anaconda/4.3.1/lib/python2.7/site-packages/nipype/interfaces/utility/wrappers.py", line 146, in _run_interface
    self.imports)
  File "/apps/software/anaconda/4.3.1/lib/python2.7/site-packages/nipype/utils/misc.py", line 102, in create_function_from_source
    raise_from(RuntimeError(msg), e)
  File "/apps/software/anaconda/4.3.1/lib/python2.7/site-packages/future/utils/__init__.py", line 454, in raise_from
    raise e
RuntimeError: 
Error executing function:
 
Functions in connection strings have to be standalone.
They cannot be declared either interactively or inside
another function or inline in the connect string. Any
imports should be done inside the function
Interface Function failed to run. 

170601-20:23:46,403 workflow INFO:
Traceback (most recent call last):
	 ***********************************
  File "/home/myanezl2/workspace/c3nl/QA_matlab/QA_Matlab.py", line 192, in <module>
170601-20:23:46,403 workflow ERROR:
	 could not run node: QA.QA_email
170601-20:23:46,404 workflow INFO:
	 crashfile: /home/myanezl2/workspace/c3nl/QA_matlab/crash-20170601-202344-myanezl2-QA_email-98e03d45-765c-4bec-87c5-c4e253113c83.pklz
170601-20:23:46,404 workflow INFO:
	 ***********************************
    QA.run('MultiProc', plugin_args={'n_procs': 2})
  File "/apps/software/anaconda/4.3.1/lib/python2.7/site-packages/nipype/pipeline/engine/workflows.py", line 590, in run
    runner.run(execgraph, updatehash=updatehash, config=self.config)
  File "/apps/software/anaconda/4.3.1/lib/python2.7/site-packages/nipype/pipeline/plugins/base.py", line 279, in run
    report_nodes_not_run(notrun)
  File "/apps/software/anaconda/4.3.1/lib/python2.7/site-packages/nipype/pipeline/plugins/base.py", line 101, in report_nodes_not_run
    raise RuntimeError(('Workflow did not execute cleanly. '
RuntimeError: Workflow did not execute cleanly. Check log for details

Process finished with exit code 1
```
### Expected behavior
This function was working in this previous installation:
print(nipype.__version__)
0.11.0

print(nipype.get_info())
{'nibabel_version': '2.0.2', 'networkx_version': '1.10', 'numpy_version': '1.11.0', 'sys_platform': 'linux2', 'sys_version': '2.7.11 |Anaconda 2.4.1 (64-bit)| (default, Dec  6 2015, 18:08:32) \n[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]', 'commit_source': '(none found)', 'commit_hash': '<not found>', 'pkg_path': '/share/apps/anaconda-3-2.4.1/lib/python2.7/site-packages/nipype', 'sys_executable': '/share/apps/anaconda-3-2.4.1/bin/python', 'traits_version': '4.5.0', 'scipy_version': '0.17.0'}

### Script/Workflow details
```
def gen_qa_email(in_directory, empty_variable, in_date):
	import os
	import smtplib
	import pdfkit
	from email.mime.application import MIMEApplication
	from email.mime.multipart import MIMEMultipart
	from email.mime.text import MIMEText
	from email.mime.image import MIMEImage

	""" Convert html to pdf"""
	pdf_file = pdfkit.from_file(in_directory + '/QA_results.html', False)

	""" Create the email message container from the input args."""
	recipients = ['email@email.com']
	msg = MIMEMultipart('alternative')
	msg['Subject'] = 'QA' + in_date
	msg['From'] = 'email@email.com'
	msg['To'] = ", ".join(recipients)

	""" Attach parts into message container"""
	pdf = MIMEApplication(pdf_file, 'pdf')
	pdf.add_header('Content-Disposition', 'attachment', filename= 'QA.pdf')
	msg.attach(pdf)

	try:
		smtpserver = smtplib.SMTP("smtp.gmail.com", 587)
		smtpserver.ehlo()
		smtpserver.starttls()
		smtpserver.ehlo
		smtpserver.login(''email@email.com']', 'password')
		smtpserver.sendmail(''email@email.com'], recipients, msg.as_string())
		smtpserver.quit()
	except smtplib.SMTPAuthenticationError as e:
		print "Unable to send message: %s" % e
	return None

QA_email = Node(name='QA_email', interface=Function(input_names=['in_directory', 'empty_variable', 'in_date'], output_names=[], function=gen_qa_email))
QA_email.inputs.in_date = experiment_date
```
### Platform details:

print(nipype.__version__)
0.13.1

print(nipype.get_info())
{'nibabel_version': '2.1.0', 'sys_executable': '/apps/software/anaconda/4.3.1/bin/python', 'networkx_version': '1.11', 'numpy_version': '1.11.3', 'sys_platform': 'linux2', 'sys_version': '2.7.13 |Anaconda 4.3.1 (64-bit)| (default, Dec 20 2016, 23:09:15) \n[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]', 'commit_source': u'installation', 'commit_hash': u'8946bcab9', 'pkg_path': '/apps/software/anaconda/4.3.1/lib/python2.7/site-packages/nipype', 'nipype_version': u'0.13.1', 'traits_version': '4.6.0', 'scipy_version': '0.18.1'}

### PR DESCRIPTION
**Pull request by mgxd**
Fixes #2060 #2038 .

Changes proposed in this pull request
- FIX: allow python 2 code in `Function` again

### CODE REVIEW

**Comment by satra** [comment]
@mgxd - this is not a good solution as we do want people to use print function everywhere. it's completely fine in python 2.

**Comment by mgxd** [comment]
@satra hm, wouldn't we want to support all python 2 legacy code though? And then once 1.0 rolls around, drop py2 support altogether

**Comment by satra** [comment]
@mgxd - i see a change but no specific test for the issue. could you please add a test?

**Comment by oesteban** [comment]
@mgxd @satra I guess the issue here is that we have the `from __future__ import ...` only at the top of files, and since functions are serialized within the `Function` interface, those imports don't work anymore.

The "right" fix here would not be adding:

```Python
from __future__ import print_function, division, unicode_literals, absolute_import
from future import standard_library
standard_library.install_aliases()
from builtins import str, bytes, range, zip
```

right after the function definition in the string?

**Comment by mgxd** [comment]
@oesteban To me, tt looks like the function wrapped in the `Function` interface does get the `from __future__` imports applied - however by including `unicode_literals`, we are forcing any external packages imported within the function to be flexible enough to handle unicode by overwritting the default `str`.

Essentially, this seems to be the chain of events
```python
# nipype/misc/utils
from __future__ import unicode_literals # and the others
# function code
import numpy as np
this_array = np.array(['1.1'])
np.savetxt('test.txt', this_array, fmt='%.6f')
```

resulting in `ValueError: invalid fmt: u'%.6f'`

Could you expand on what you meant be the "right" fix?

**Comment by oesteban** [comment]
Well, I meant the "right" fix because `Function` interfaces are 2-3 incompatible islands at the moment. My suggestion would indeed require any wrapped function to be python 3 compatible.

File: nipype/utils/tests/test_functions.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    def print_statement():
+        # test python 2 compatibility
+        exec('print ""')
+
+    wrapped_func = create_function_from_source(getsource(is_string))
+    assert is_string() == wrapped_func()
+
+    wrapped_func2 = create_function_from_source(getsource(print_statement))
+    wrapped_func2()
```
Comment (satra) [inline nipype/utils/tests/test_functions.py]: no test here?

Comment (mgxd) [inline nipype/utils/tests/test_functions.py]: will raise syntax error if print can't be done - is there a better way to do this?

Comment (satra) [inline nipype/utils/tests/test_functions.py]: with pytest.raises(SyntaxError): wrapped_func2()

Comment (satra) [inline nipype/utils/tests/test_functions.py]: ah but it doesn't error by default. so you should return something from the function call and check with assert.

File: nipype/utils/tests/test_functions.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    try:
+        exec('print ""')
+        return True
+    except SyntaxError:
+        return False
+
+@pytest.mark.skipif(sys.version_info[0] > 2, reason="breaks python 3")
+def test_func_py2():
+    def is_string():
+        return isinstance('string', str)
```
Comment (satra) [inline nipype/utils/tests/test_functions.py]: why would this function break python 3?

Comment (satra) [inline nipype/utils/tests/test_functions.py]: perhaps separate out into two tests - the print statement which breaks python 3 and the function which should be fine for both.

Comment (mgxd) [inline nipype/utils/tests/test_functions.py]: @satra it'll raise an `AssertionError` - I've isolated the tests though

**Pull request merged**